test/language/expressions/optional-chaining/iteration-statement-for-of-type-error.js

```diff
@@ -0,0 +1,18 @@
+// Copyright 2019 Google, LLC.  All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: prod-OptionalExpression
+description: >
+  optional chain returning undefined in RHS of for of statement
+info: |
+  IterationStatement
+    for (LeftHandSideExpression of Expression) Statement
+features: [optional-chaining]
+---*/
+
+assert.throws(TypeError, function() {
+  const obj = undefined;
+  for (const key of obj?.a) {
+    str += key;
```

Comment:
this line is not necessary.

To avoid any sort of false positives, I'd suggest splitting in 4 simple statements:

```js
assert.throws(TypeError, function() {
  for (const key of {}?.a) ;
});

assert.throws(TypeError, function() {
  for (const key of {}?.a) {}
});

const obj = undefined; // outside of the error catching.
assert.throws(TypeError, function() {
  for (const key of obj?.a) {}
});

assert.throws(TypeError, function() {
  for (const key of obj?.a) ;
});
```